### PR TITLE
chore: mls documentation in data model - WPB-10272

### PR DESCRIPTION
**Pull request by David-Henner**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-10272" title="WPB-10272" target="_blank">Sub-taskWPB-10272</a>  [iOS] MLS DocC: Document MLS APIs in data model
  </td></table>
  <br />
 

This PR adds documentation for MLS in data model. It introduces DocC articles and DocC comments in key MLS classes.

To consult the documentation: checkout the branch in Xcode and run `Product > Build Documentation`. It will build and open the DocC documentation for the project. You can then navigate to `WireDataModel` and you'll find sections for MLS, MLS migration, E2EI and Core Crypto. There are articles introducing MLS and the MLS migration. E2EI and Core Crypto is set up to organise the files but does not contain articles or extensive comments yet.

---

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [ ] Adds/updates automated tests.

### CODE REVIEW

File: wire-ios-data-model/Source/MLS/MLSDecryptionService.swift [hunk 1]
```diff
@@ -24,15 +24,44 @@ import WireSystem
...
+    /// In addition to decrypting the message and returning a result, this method will also publish events if the epoch has changed
+    /// or if new CRL distribution points have been found.
+
     func decrypt(
         message: String,
         for groupID: MLSGroupID,
         subconversationType: SubgroupType?
     ) async throws -> [MLSDecryptResult]
 
+
```
Comment (David-Henner) [inline wire-ios-data-model/Source/MLS/MLSDecryptionService.swift]:
```suggestion
```

**Comment by jullianm** [comment]
Looking good! Nice to have that built-in documentation

**Review by jullianm — APPROVED** [review]

File: wire-ios-data-model/WireDataModel.docc/MLS.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+- **Ciphersuite**: A specific set of cryptographic algorithms used to secure communications within a group. It defines the algorithms for key exchange, encryption, message authentication, and hashing, ensuring that all participants in the group use the same cryptographic methods.
+- **Key package**: A data structure that contains a public key, supported ciphersuites, and other necessary cryptographic material for a client. It is used during the initial setup of a group or when a new member joins the group.
+
+## Core crypto
+
+Core crypto is a wrapper on top of OpenMLS aimed to provide an ergonomic API to create, manage and interact with MLS groups. It abstracts MLS and Proteus. 
+
+The library is written in Rust and provides Swift bindings generated with FFI. The dependency is distributed as an `xcframework` and is managed with Carthage.
+
+[documentation](https://wireapp.github.io/core-crypto/core_crypto/index.html)
```
Comment (samwyndham) [inline wire-ios-data-model/WireDataModel.docc/MLS.md]:
```suggestion
The library is written in Rust and provides Swift bindings generated with FFI. The dependency is distributed as an `xcframework` and is managed with Carthage. [See documentation](https://wireapp.github.io/core-crypto/core_crypto/index.html)
```

Comment (samwyndham) [reply wire-ios-data-model/WireDataModel.docc/MLS.md]: The `documentation` link by itself is a bit confusing when rendered

File: wire-ios-data-model/WireDataModel.docc/MLS.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+## MLS Groups / Conversations
+
+MLS conversations are referred through the codebase as "conversation" or "group". These terms are often used interchangeably but their origins differ: one comes from how conversations are represented in Core Data, and the other comes from Core Crypto.
+
+In Core Data, conversations are represented by ``ZMConversation`` objects. They contain metadata about the conversations such as their remote identifier, domain, name, etc... 
+
+They're used to represent both MLS and Proteus conversations. The protocol they use is represented by the ``ZMConversation/messageProtocol`` property.
+
+``ZMConversation`` objets have a couple of optionnal properties that are specific to MLS conversations, such as:
```
Comment (samwyndham) [inline wire-ios-data-model/WireDataModel.docc/MLS.md]:
```suggestion
``ZMConversation`` objets have a couple of optional properties that are specific to MLS conversations, such as:
```

File: wire-ios-data-model/WireDataModel.docc/MLS Migration.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# MLS Migration
+
+The MLS migration is about migrating existing proteus conversations to MLS. The group conversations are migrated following one strategy, whereas the 1:1 conversations are migrated following another.
```
Comment (samwyndham) [inline wire-ios-data-model/WireDataModel.docc/MLS Migration.md]:
```suggestion
MLS migration is about migrating existing proteus conversations to MLS. Group conversations are migrated following one strategy, whereas 1:1 conversations are migrated following another.
```

File: wire-ios-data-model/WireDataModel.docc/MLS Migration.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+## Group conversations
+
+For group conversations, the migration is designed to be done over a period of time defined by a starting time and an optional finalization time. During this period, the conversations transitions to the ``MessageProtocol/mixed`` protocol until they are ready to be fully migrated to ``MessageProtocol/mls``.
+
+Conversations in `.mixed` mode use Proteus to encrypt and decrypt message, and retain the capacity to add and remove members from the Proteus part of the conversatin, whilst also maintaining their MLS group counterpart. Meaning that they'll support operations such as committing pending proposals and updating key material.
+Once every member of the conversation is capable of supporting MLS, or once the finalization time is reached, the conversation is fully migrated to MLS.
+
+The configuration of the migration is defined by the backend and is fetched by the client. The client is responsible for starting the migration at the right time and for finalizing it when the time comes. The configuration is defined in ``Feature/MLSMigration``
+
+The ``ProteusToMLSMigrationCoordinator`` class is responsible for managing the migration of group conversation.
```
Comment (samwyndham) [inline wire-ios-data-model/WireDataModel.docc/MLS Migration.md]:
```suggestion
The ``ProteusToMLSMigrationCoordinator`` class is responsible for managing the migration of group conversations.
```

File: wire-ios-data-model/WireDataModel.docc/MLS Migration.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# MLS Migration
+
+The MLS migration is about migrating existing proteus conversations to MLS. The group conversations are migrated following one strategy, whereas the 1:1 conversations are migrated following another.
+
+## Group conversations
+
+For group conversations, the migration is designed to be done over a period of time defined by a starting time and an optional finalization time. During this period, the conversations transitions to the ``MessageProtocol/mixed`` protocol until they are ready to be fully migrated to ``MessageProtocol/mls``.
```
Comment (samwyndham) [inline wire-ios-data-model/WireDataModel.docc/MLS Migration.md]:
```suggestion
For group conversations, the migration is designed to be done over a period of time defined by a starting time and an optional finalization time. During this period, the conversations transition to the ``MessageProtocol/mixed`` protocol until they are ready to be fully migrated to the ``MessageProtocol/mls`` protocol.
```

File: wire-ios-data-model/WireDataModel.docc/MLS Migration.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+Conversations in `.mixed` mode use Proteus to encrypt and decrypt message, and retain the capacity to add and remove members from the Proteus part of the conversatin, whilst also maintaining their MLS group counterpart. Meaning that they'll support operations such as committing pending proposals and updating key material.
```
Comment (samwyndham) [inline wire-ios-data-model/WireDataModel.docc/MLS Migration.md]:
```suggestion
Conversations in `.mixed` mode use Proteus to encrypt and decrypt messages, and retain the capacity to add and remove members from the Proteus part of the conversation, whilst also maintaining their MLS group counterpart. This means that they support operations such as committing pending proposals and updating key material.
```

File: wire-ios-data-model/WireDataModel.docc/MLS Migration.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+Once every member of the conversation is capable of supporting MLS, or once the finalization time is reached, the conversation is fully migrated to MLS.
+
+The configuration of the migration is defined by the backend and is fetched by the client. The client is responsible for starting the migration at the right time and for finalizing it when the time comes. The configuration is defined in ``Feature/MLSMigration``
+
+The ``ProteusToMLSMigrationCoordinator`` class is responsible for managing the migration of group conversation.
+
+Documentation about implementation details is available on [Confluence](https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/746488003/Proteus+to+MLS+Migration).
```
Comment (samwyndham) [inline wire-ios-data-model/WireDataModel.docc/MLS Migration.md]:
```suggestion
Documentation about the implementation details are available on [Confluence](https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/746488003/Proteus+to+MLS+Migration).
```

File: wire-ios-data-model/Source/MLS/CommitSender.swift [hunk 9]
```diff
@@ -23,20 +23,52 @@ import WireCoreCrypto
...
         for groupID: MLSGroupID
     ) async throws -> [ZMUpdateEvent]
 
+    /// Returns a publisher that emits the group ID of the group when the epoch changes.
+
     func onEpochChanged() -> AnyPublisher<MLSGroupID, Never>
 
 }
 
+/// This class is responsible for sending commits and external commits and handling the results.
```
Comment (samwyndham) [inline wire-ios-data-model/Source/MLS/CommitSender.swift]:
```suggestion
/// An actor responsible for sending commits and external commits and handling the results.
```

File: wire-ios-data-model/Source/MLS/StaleMLSKeyMaterialDetector.swift [hunk 10]
```diff
@@ -38,6 +38,9 @@ public protocol StaleMLSKeyDetectorProtocol {
 
 }
 
+/// This class is responsible for keeping track of groups that have stale key material.
```
Comment (samwyndham) [inline wire-ios-data-model/Source/MLS/StaleMLSKeyMaterialDetector.swift]:
```suggestion
/// A class responsible for keeping track of groups that have stale key material.
```

File: wire-ios-data-model/Source/MLS/Migration/ProteusToMLSMigrationCoordinator.swift [hunk 11]
```diff
@@ -22,9 +22,17 @@ import WireUtilities
 
 // sourcery: AutoMockable
 public protocol ProteusToMLSMigrationCoordinating {
+
+    /// Updates the migration status of group conversations.
     func updateMigrationStatus() async throws
+
 }
 
+/// This class is responsible for migrating group conversations from Proteus to MLS.
```
Comment (samwyndham) [inline wire-ios-data-model/Source/MLS/Migration/ProteusToMLSMigrationCoordinator.swift]:
```suggestion
/// A class responsible for migrating group conversations from Proteus to MLS.
```

File: wire-ios-data-model/Source/MLS/MLSEncryptionService.swift [hunk 12]
```diff
@@ -22,13 +22,27 @@ import WireCoreCrypto
...
+    ///   A byte array representing the ciphertext.
+    
     func encrypt(
         message: Data,
         for groupID: MLSGroupID
     ) async throws -> Data
 
 }
 
+/// This class is responsible for encrypting messages for MLS groups.
```
Comment (samwyndham) [inline wire-ios-data-model/Source/MLS/MLSEncryptionService.swift]:
```suggestion
/// A class responsible for encrypting messages for MLS groups.
```

File: wire-ios-data-model/Source/MLS/MLSEncryptionService.swift [hunk 13]
```diff
@@ -22,13 +22,27 @@ import WireCoreCrypto
...
+    /// - Parameters:
+    ///   - message: data representing the plaintext message
+    ///   - groupID: the id of the MLS group in which to encrypt
+    ///
+    /// - Throws:
+    ///   `MLSMessageEncryptionError` if the message couldn't be encrypted
+    ///
+    /// - Returns:
+    ///   A byte array representing the ciphertext.
+
```
Comment (David-Henner) [inline wire-ios-data-model/Source/MLS/MLSEncryptionService.swift:36]:
```suggestion

```

File: wire-ios-data-model/Source/MLS/MLSActionExecutor.swift [hunk 14]
```diff
@@ -22,18 +22,94 @@ import WireCoreCrypto
 
 public protocol MLSActionExecutorProtocol {
 
+    /// Processes a welcome message.
+    ///
+    /// - Parameter message: The welcome message to process
+    /// - Returns: The group ID of the group the welcome message was for
+    ///
+    /// If any new CRL distribution points are found, they will be published. They can be observed with ``MLSActionExecutor/onNewCRLsDistributionPoints()``
```
Comment (samwyndham) [inline wire-ios-data-model/Source/MLS/MLSActionExecutor.swift]: Optional - I don't think you have much time: Consider hard wrapping this and other doc comment at 120 lines. They seem to render correctly (still the same) once documentation is built.

File: wire-ios-data-model/Source/MLS/MLSActionExecutor.swift [hunk 15]
```diff
@@ -22,18 +22,94 @@ import WireCoreCrypto
 
 public protocol MLSActionExecutorProtocol {
 
+    /// Processes a welcome message.
+    ///
+    /// - Parameter message: The welcome message to process
```
Comment (samwyndham) [inline wire-ios-data-model/Source/MLS/MLSActionExecutor.swift]: Missing periods at end of sentences compared to other doc comments.

File: wire-ios-data-model/Source/MLS/MLSActionExecutor.swift [hunk 16]
```diff
@@ -22,18 +22,94 @@ import WireCoreCrypto
...
+
     func onEpochChanged() -> AnyPublisher<MLSGroupID, Never>
+
+    /// Returns a publisher that emits the new CRL distribution points when they are found
+
     func onNewCRLsDistributionPoints() -> AnyPublisher<CRLsDistributionPoints, Never>
 
 }
 
+/// This class is responsible for performing commits on MLS groups and decrypting messages in a non-reentrant manner.
```
Comment (samwyndham) [inline wire-ios-data-model/Source/MLS/MLSActionExecutor.swift]:
```suggestion
/// An actor responsible for performing commits on MLS groups and decrypting messages in a non-reentrant manner.
```

File: wire-ios-data-model/Source/MLS/MLSDecryptionService.swift [hunk 17]
```diff
@@ -24,15 +24,43 @@ import WireSystem
...
+    /// Publishes an event when new CRL distribution points are found.
+
     func onNewCRLsDistributionPoints() -> AnyPublisher<CRLsDistributionPoints, Never>
 
+    /// Decrypts an MLS message for the given group
+    ///
+    /// - Parameters:
+    ///   - message: a base64 encoded encrypted message
+    ///   - groupID: the id of the MLS group
+    ///   - subconversationType: the type of subconversation (if it exists)  to which this message belongs
```
Comment (samwyndham) [inline wire-ios-data-model/Source/MLS/MLSDecryptionService.swift]:
```suggestion
    ///   - subconversationType: the type of subconversation (if it exists) to which this message belongs
```

File: wire-ios-data-model/Source/MLS/MLSDecryptionService.swift [hunk 18]
```diff
@@ -61,6 +89,9 @@ protocol DecryptedMessageBundle {
 extension DecryptedMessage: DecryptedMessageBundle { }
 extension BufferedDecryptedMessage: DecryptedMessageBundle { }
 
+/// This class is responsible for decrypting messages for MLS groups.
```
Comment (samwyndham) [inline wire-ios-data-model/Source/MLS/MLSDecryptionService.swift]:
```suggestion
/// A class responsible for decrypting messages for MLS groups.
```

File: wire-ios-data-model/Source/MLS/SubconversationGroupIDRepository.swift [hunk 19]
```diff
@@ -21,22 +21,49 @@ import Foundation
...
+    ///
+    /// **Note:** This will return `nil` if the group ID hasn't previously been stored by calling ``storeSubconversationGroupID(_:forType:parentGroupID:)``
+
     func findSubgroupTypeAndParentID(
         for targetGroupID: MLSGroupID
     ) async -> (parentID: MLSGroupID, type: SubgroupType)?
 }
 
+
+/// This class is responsible for storing and fetching subconversations group IDs.
```
Comment (samwyndham) [inline wire-ios-data-model/Source/MLS/SubconversationGroupIDRepository.swift]:
```suggestion
/// An actor responsible for storing and fetching subconversations group IDs.
```

**Review by samwyndham — APPROVED** [review]
Nice work! I added a bunch of picky comments. Feel free to ignore any of them. They are mostly opinionated rather than objective

File: wire-ios-data-model/Source/MLS/MLSActionExecutor.swift (hunk 15 above)
Comment (David-Henner) [inline wire-ios-data-model/Source/MLS/MLSActionExecutor.swift]:
```suggestion
    /// - Parameter message: The welcome message to process.
```

File: wire-ios-data-model/Source/MLS/MLSActionExecutor.swift [hunk 20]
```diff
@@ -22,18 +22,94 @@ import WireCoreCrypto
 
 public protocol MLSActionExecutorProtocol {
 
+    /// Processes a welcome message.
+    ///
+    /// - Parameter message: The welcome message to process
+    /// - Returns: The group ID of the group the welcome message was for
```
Comment (David-Henner) [inline wire-ios-data-model/Source/MLS/MLSActionExecutor.swift]:
```suggestion
    /// - Returns: The group ID of the group the welcome message was for.
```

File: wire-ios-data-model/Source/MLS/MLSActionExecutor.swift (hunk 14 above)
Comment (David-Henner) [reply wire-ios-data-model/Source/MLS/MLSActionExecutor.swift]: do you mean 120 characters per line?

File: wire-ios-data-model/Source/MLS/MLSEncryptionService.swift (hunk 13 above)
Comment (David-Henner) [inline wire-ios-data-model/Source/MLS/MLSEncryptionService.swift]:
```suggestion

```

File: wire-ios-data-model/Source/MLS/MLSService.swift [hunk 21]
```diff
@@ -23,49 +23,233 @@ import WireCoreCrypto
...
+    /// Calls ``MLSService/createGroup(for:parentGroupID:)`` to create the group.
+    ///
+    /// Then calls ``MLSService/addMembersToConversation(with:for:)`` to add the self user's clients to the group.
+    ///
+    /// If it catches ``MLSService/MLSAddMembersError/noMembersToAdd`` it will update the key material to send a commit to the backend and advance to a new epoch.
+    ///
+    /// [confluence use case](https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/791412993/Use+case+join+self+conversation+MLS)
 
     func createSelfGroup(for groupID: MLSGroupID) async throws -> MLSCipherSuite
+
```
Comment (David-Henner) [inline wire-ios-data-model/Source/MLS/MLSService.swift]:
```suggestion

```

File: wire-ios-data-model/Source/MLS/MLSService.swift [hunk 22]
```diff
@@ -23,49 +23,233 @@ import WireCoreCrypto
...
+    ///   - parentQualifiedID: The qualified ID of the parent
+    ///   - parentGroupID: The group ID of the parent
+    ///   - subconversationType: The subgroup type
+    ///
+    /// Fetches the subgroup ID based on the type and parent group IDs.
+    ///
+    /// It will request the backend to leave the subgroup, and if successful,
+    /// remove the subgroup from the ``SubconversationGroupIDRepository``
+    /// and from Core Crypto's local storage.
+
```
Comment (David-Henner) [inline wire-ios-data-model/Source/MLS/MLSService.swift]:
```suggestion

```

**Comment by David-Henner** [comment]
Thank you @samwyndham for spotting all the typos!

File: wire-ios-data-model/Source/MLS/MLSActionExecutor.swift (hunk 14 above)
Comment (David-Henner) [reply wire-ios-data-model/Source/MLS/MLSActionExecutor.swift]: fixed https://github.com/wireapp/wire-ios/pull/1857/commits/0a6590b72c17249803104e54458a90b00d8c47ce

**Pull request merged**